### PR TITLE
Fix: stutter avatar on moving platform

### LINKED ISSUE #1029: [QA] Avatar positition stutters when moving on platforms
The avatar shakes up and down when displacing on moving platforms as resolves collisions. Is probable to fall through the platform if the speed is too high.

The character should be parented to the platforms to inherit position?

### LINKED ISSUE #1068: [QA] Partially Hidden Footwear
### **Explorer Alpha build version:**
0.3.383-g852675f1-main

### **Issue Description:**
In the Decentraland Explorer Alpha app, the platform where the avatar stands, both in the login screen and in the backpack, partially hides the avatar's footwear, giving the impression that it is slightly buried.

### **Steps to reproduce:**

1. Launch Decentraland EA with any account.
2. Observe the avatar's footwear on the login screen.
3. Log in.
4. Observe the avatar's footwear in the backpack.

### **Expected behavior:**
The platform should not obscure any part of the avatar's footwear, providing a clear and accurate representation of the avatar's appearance on both the login screen and in the backpack.

### **Current behavior:**
The platform where the avatar stands obscures part of the footwear, making it appear as if the avatar's feet are partially buried, both on the login screen and in the backpack.

13 05 2024_11 14 42_REC
13 05 2024_11 16 05_REC

### **Severity:**

- [ ] Cosmetic
- [] Minor
- [❌ ] Medium
- [ ] High
- [ ] Critical
- [ ] Blocker

### **Priority:**
- [ ] High
- [❌ ] Medium
- [ ] Low

### **Platforms / OS:**
- [❌ ] Windows 11
- [ ] Windows 10
- [ ] Mac Sonoma
- [ ] Mac (specify the version on Additional Notes)

#### **Additional Notes:**
Windows 11. 12th Gen Intel(R) Core(TM) i7-12650H   2.30 GHz

### LINKED ISSUE #1294: goerli: Rocket-Board: Avatar Falls Off Moving Platform
### **Explorer Alpha build version:**

- **v0.4.0**
- **v0.10.0**
- **v0.11.0**
- **v0.12.0**
- **v0.13.0**
- **v0.14.0**

### **Issue Description:**
In the Goerli scene Rocket-Board (72,1), the avatar falls off the platform while it is in motion. 

### **Steps to reproduce:**

1. /goto goerli
2. Navigate to the Rocket-Board scene (72,1)
3. Stand on the platform.
4. Use E and F to control flying platform

### **Expected behavior:**

https://github.com/decentraland/unity-explorer/assets/146741383/de632bd4-a9f6-4d88-bd1a-01882231d3e7

### **Current behavior:**

https://github.com/decentraland/unity-explorer/assets/146741383/5217cb61-4500-4236-8fbb-b193ce0de493

🖥️ Platform
- Windows 11. 12th Gen Intel(R) Core(TM) i7-12650H 2.30 GHz. nvidia geforce RTX
- Macbook M1 Pro Sonoma 14.5

### PR DESCRIPTION
**Pull request by popuz**
## What does this PR change?
fix #1294 
fix #1029 
fix #1068

- disable gravity when player is on moving platform

## How to test the changes?

1. Check bug tickets for repro steps
2. Verify other good movement is not changed

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: Explorer/Assets/DCL/Character/CharacterMotion/Platforms/PlatformRaycast.cs [hunk 1]
```diff
@@ -9,30 +10,31 @@ namespace DCL.CharacterMotion.Platforms
...
-                    platformComponent.LastAvatarRelativePosition = platformComponent.CurrentPlatform.InverseTransformPoint(transform.position);
-                    platformComponent.LastAvatarRelativeRotation = platformComponent.CurrentPlatform.InverseTransformDirection(transform.forward);
-                }
+                platformComponent.PlatformCollider = hitInfo.collider;
+                platformComponent.CurrentPlatform = hitInfo.collider.transform;
+
+                platformComponent.LastPlatformPosition = hitInfo.collider.transform.position;
+                platformComponent.LastPlatformPosition = null;
+                platformComponent.LastAvatarRelativePosition = platformComponent.CurrentPlatform.InverseTransformPoint(characterTransform.position);
+                platformComponent.LastAvatarRelativeRotation = platformComponent.CurrentPlatform.InverseTransformDirection(characterTransform.forward);
```
Comment (AlejandroAlvarezMelucciDCL) [inline Explorer/Assets/DCL/Character/CharacterMotion/Platforms/PlatformRaycast.cs:36]: Is this intended? You are assigning the value for `platformComponent.LastPlatformPosition` twice and the second is `null`, maybe we could remove line 33?

File: Explorer/Assets/DCL/Utilities/DCLPhysics.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    /// </summary>
+    public static class DCLPhysics
+    {
+        /// <summary>
+        /// Unity SphereCast with Debug gizmos.
+        /// </summary>
+        public static bool SphereCast(Ray ray, float radius, out RaycastHit hitInfo, float maxDistance, int layerMask, QueryTriggerInteraction queryTriggerInteraction)
+        {
+            bool hasHit = Physics.SphereCast(ray, radius, out hitInfo, maxDistance, layerMask, queryTriggerInteraction);
+            // if (Application.isEditor) DebugUtils.DrawRaycast(radius, ray.origin, maxDistance, hasHit, hitInfo);
```
Comment (AlejandroAlvarezMelucciDCL) [inline Explorer/Assets/DCL/Utilities/DCLPhysics.cs:16]: Did you mean to leave this commented out?

File: Explorer/Assets/DCL/Utilities/DebugUtils.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+﻿using System.Diagnostics;
+using UnityEngine;
+using Debug = UnityEngine.Debug;
+
+namespace DCL.Utilities
+{
+    public class DebugUtils
+    {
+        [Conditional("UNITY_EDITOR")]
+        public static void DrawRaycast(float radius, Vector3 rayOrigin, float rayDistance, bool hasHit, RaycastHit hitInfo)
```
Comment (AlejandroAlvarezMelucciDCL) [inline Explorer/Assets/DCL/Utilities/DebugUtils.cs:10]: I think it's a bit weird to have the `radius` as first param 🤔 What do you think about moving it after `rayDstance`?

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/InterpolateCharacterSystem.cs [hunk 4]
```diff
@@ -95,13 +103,26 @@ private void TeleportPlayer(in Entity entity, in CharacterController controller,
...
+            PlatformSaveLocalPosition.Execute(ref platformComponent, characterTransform.position);
 
             // In order to detect if we got stuck between 2 slopes we just check if our vertical delta movement is zero when on a slope
             if (rigidTransform.IsOnASteepSlope && Mathf.Abs(deltaMovement.sqrMagnitude) <= ALMOST_ZERO)
                 rigidTransform.IsStuck = true;
             else
                 rigidTransform.IsStuck = false;
         }
+
+        private static Vector3 GravityDelta(float dt, CharacterRigidTransform rigidTransform, CharacterPlatformComponent platformComponent)
```
Comment (AlejandroAlvarezMelucciDCL) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/InterpolateCharacterSystem.cs:115]: Shall we rename this to `GetGravityDelta`? 🤔 WDYT

**Review by AlejandroAlvarezMelucciDCL — DISMISSED** [review]
Looks good overall! Just left some minor comments :)

File: Explorer/Assets/DCL/Character/CharacterMotion/Platforms/PlatformRaycast.cs (hunk 1 above)
Comment (popuz) [reply Explorer/Assets/DCL/Character/CharacterMotion/Platforms/PlatformRaycast.cs:36]: good catch 🎣

File: Explorer/Assets/DCL/Utilities/DebugUtils.cs (hunk 3 above)
Comment (popuz) [reply Explorer/Assets/DCL/Utilities/DebugUtils.cs:10]: especially having radius for the raycast 😅 could be quite confusing, right

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/InterpolateCharacterSystem.cs (hunk 4 above)
Comment (popuz) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/InterpolateCharacterSystem.cs:115]: I was thinking to use it directly in equation and omitting the variable, but the variable is used several time. So I will call it `CalculateGravityDelta`

**Review by AlejandroAlvarezMelucciDCL — APPROVED** [review]
LGTM !

**Pull request merged**